[schema registry json] Initialize schema registry json library (#26673)
Azure/azure-sdk-for-js: sdk/schemaregistry/schema-registry-json/README.md

Azure Schema Registry is a schema repository service hosted by Azure Event Hubs,
providing schema storage, versioning, and management. This package provides an
Json serializer capable of serializing and deserializing payloads containing
Json-serialized data.

Key links:

- [Source code](https://github.com/Azure/azure-sdk-for-js/tree/schemaregistryjson-init/sdk/schemaregistry/schema-registry-json)
- [Package (npm)](https://www.npmjs.com/package/@azure/schema-registry-json)
- [API Reference Documentation](https://aka.ms/schemaregistryjson-js-api)
- [Samples](https://github.com/Azure/azure-sdk-for-js/tree/schemaregistryjson-init/sdk/schemaregistry/schema-registry-json/samples)

## Getting started

- [LTS versions of Node.js](https://github.com/nodejs/release#release-schedule)

### Prerequisites

- An [Azure subscription][azure_sub]
- An existing [Schema Registry resource](https://aka.ms/schemaregistry)

### Install the `@azure/schema-registry-json` package

Install the Azure Text Analytics client library for JavaScript with `npm`:

```bash
npm install @azure/schema-registry-json
```

## Key concepts

### JsonSerializer

> Provides API to serialize to and deserialize from Json Binary Encoding wrapped in a message

Reviewer: ```suggestion
Provides API to serialize to and deserialize from JSON wrapped in a message
```

with a content type field containing the schema ID. Uses
`SchemaRegistryClient` from the [@azure/schema-registry](https://www.npmjs.com/package/@azure/schema-registry) package
to get schema IDs from schema definition or vice versa. The provided API has internal cache to avoid calling the schema registry service when possible.

### Messages

By default, the serializer will create messages structured as follows:

> - `data`: a byte array containing data in the Json Binary Encoding. Note that it
>   is NOT Json Object Container File. The latter includes the schema and creating
>   it defeats the purpose of using this serializer to move the schema out of the
>   message payload and into the schema registry.

Reviewer: ```suggestion
- `data`: a byte array containing JSON data.
```

  "<fully qualified namespace>",
  new DefaultAzureCredential()
);
const serializer = new JsonSerializer(client, {
  groupName: "<group>",
  messageAdapter: createEventDataAdapter(),
});

// Example Json schema
const schema = JSON.stringify({  
  $schema: "http://json-schema.org/draft-04/schema#",
  $id: "person",
  title: "Student",
  description: "A student in the class",
  type: "object",
  properties: {
    name: {
      type: "string",
      description: "The name of the student",
    },
  },
  required: ["name"]
});

// Example value that matches the Json schema above
const value = { name: "Bob" };

// Serialize value to a message
const message = await serializer.serialize(value, schema);

// Deserialize a message to value
const deserializedValue = await serializer.deserialize(message);
```

Validation may be useful for your application since JSON deserialization is flexible. 
Deserializing on its own will not verify that the value matches a given schema, 
so implementing the validation method will be a much more reliable way to enforce schema adherance.
To see how the validation might be implemented, please take a look at the `schemaRegistryJsonWithValidation.ts` 

> sample [here](https://github.com/Azure/azure-sdk-for-js/tree/schemaregistryjson-init/sdk/schemaregistry/schema-registry-json/samples-dev/schemaRegistryJsonWithValidation.ts)

Reviewer: I gave feedback on this part a while ago, could you please apply it?

## Troubleshooting

> The Json serializer communicates with the [Schema Registry][schema_registry] service as needed to register or query schemas and those service calls could throw a [RestError][resterror]. Furthermore, errors of type `Error` will be thrown when serialization or deserialization fails. The `cause` property will contain the underlying error that was thrown from the Json implementation library.

Reviewer: ```suggestion
The Json serializer communicates with the [Schema Registry][schema_registry] service as needed to register or query schemas and those service calls could throw a [RestError][resterror]. Furthermore, errors of type `Error` will be thrown when serialization or deserialization fails. The `cause` property will contain the underlying error that was thrown from the JSON parser.
```
